crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -970,6 +977,18 @@ where
                 let pre_loop = self.flow_snapshot();
                 let constraint = self.record_expression_constraint(test);
 
+                // We need multiple copies of the visibility constraint for the while condition,
+                // since we need to model situations where the first evaluation of the condition
+                // returns True, but a later evaluation returns False.
+                let first_vis_constraint_id =
+                    self.add_visibility_constraint(VisibilityConstraint::VisibleIf(
+                        VisibilityConstraintAtom::new_copy(constraint, 0),
+                    ));
+                let later_vis_constraint_id =
+                    self.add_visibility_constraint(VisibilityConstraint::VisibleIf(
+                        VisibilityConstraintAtom::new_copy(constraint, 1),
+                    ));
```

Comment:
I wonder if this could be a bit more ergonomic, and not have to expose `VisibilityConstraintAtom` and the implementation detail of its internal `u8` to the semantic index builder? But I see there are complexities there, and with only one call site at the moment I don't think it matters much.

Reply:
Done. I removed the `Atom` wrapper.  My thought here was to reduce the number of `, 0`s I'd have to add, but most of those were already hidden behind helper methods

Reply:
(Making the `u8` not visible at all would be more invasive, since that's how the caller up in the semantic index builder specifies that it wants distinct copies.  It's up to the caller to choose the values, such that distinct `u8`s refer to different evaluations of the constraint.  Do you have thoughts on a different API shape to express that?)

Reply:
I think I was vaguely envisioning some kind of `copy` API that would take an existing constraint ID and bump the counter behind the scenes. But the awkwardness there is that a constraint ID might not represent a `VisibleIf` constraint at all, making the API error prone. So I'm not sure that would be an improvement. It's already much nicer without the extra `Atom` type, I think what you have is good.